decryptor/base/proxy.go

```diff
@@ -92,9 +93,41 @@ type ClientSession interface {
 	Context() context.Context
 	ClientConnection() net.Conn
 	DatabaseConnection() net.Conn
+
+	PreparedStatementRegistry() PreparedStatementRegistry
+	SetPreparedStatementRegistry(registry PreparedStatementRegistry)
+	CursorRegistry() CursorRegistry
+	SetCursorRegistry(registry CursorRegistry)
 }
 
 // ProxyFactory create new Proxy for specific database
 type ProxyFactory interface {
 	New(clientID []byte, clientSession ClientSession) (Proxy, error)
 }
+
+// PreparedStatementRegistry keeps track of active prepared statements within a ClientSession.
+type PreparedStatementRegistry interface {
+	Add(statement PreparedStatement) (bool, error)
+	StatementByName(name string) (PreparedStatement, error)
```

Comment:
actually registry should allow to remove statements too because user can drop existing statements and re-assign queries by statement name. So `Add` operation should update statement if it exists or we should add one more method for updates.

Reply:
> user can drop existing statements and re-assign queries by statement name

Yeah, its possible. However, since it's not strictly necessary for correctness of Acra's processing (database will fail any attempt to use removed statements), removal tracking won't be implemented right away. Though, I do feel that we'd better have it to avoid accumulating named prepared statements, despite the expectation that their number is likely to be limited.

> So Add operation should update statement if it exists or we should add one more method for updates.

In fact, it currently acts more like an `Update`: if there is no statement then it's added to the registry, and if there is then then old one is replaced with the new one. Do you feel that we'd better name this method `Update` instead?

(My idea for the future was to have separate `Add` and `Delete` methods, with `Add` *not* allowing updates. That's because PostgreSQL protocol does not allow to update a statement in place, except for the unnamed statements. Named ones should be first removed, and then re-added under the same name.)

Reply:
> removal tracking won't be implemented right away

but why? we can track `DEALLOCATE` commands and remove prepared statement from registry on each one.
But yes, if user will deallocate and try to use deallocated prepared statement, database should return error.
But on other side each such call of prepared statement may cause some actions and calculations from acra-server. And there may be case when client's app has mistake and they use incorectly dropped prepared statements. And if it did without acra-server, they just took errors in logs and work in this state some time.
But with acra-server it may cause performance degradation if app incorrectly use and it cause extra-calculations which may slow other correct queries (for example fetching prepared statement state from distributed storage, fetching keys, censor computations, etc)

> (My idea for the future was to have separate Add and Delete methods, with Add not allowing updates. That's because PostgreSQL protocol does not allow to update a statement in place, except for the unnamed statements. Named ones should be first removed, and then re-added under the same name.)

Probably will be enough to leave Add with update logic (because of unnamed statements) but in implementation of registry we will add check that update allowable only for unnamed statement. If it is not unnamend then return error on try to add statement without `delete`.

Reply:
> but why? we can track DEALLOCATE commands and remove prepared statement from registry on each one.

By "later" I mean "probably not in the same PR if it turns out to be nontrivial", not "next year". We'll track removals. The arguments you lay out are the reason. It might not affect correctness (i.e., data observed by the application) but it has the potential to affect performance by consuming compute resources on requests which are bound to fail in the database, by consuming memory resources for effectively dead statements, etc.

I'l add removal methods here in this PR, and will keep the `Add` method with its current logic (always update), but add a TODO to refuse updates for existing named statements when we're sure that `Delete` is going to be executed as required.